src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -55,22 +51,34 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
 
                 state.Current.CollectionPropertyInitialized = true;
             }
-            else
+            else if (state.Current.IsProcessingObject(ClassType.Object))
             {
+                if (classInfo.CreateObject is null)
+                {
+                    if (classInfo.Type.IsInterface)
+                    {
+                        ThrowHelper.ThrowNotSupportedException_DeserializePolymorphicInterface(classInfo.Type);
+                    }
+                    else
+                    {
+                        ThrowHelper.ThrowNotSupportedException_DeserializeMissingParameterlessConstructor(classInfo.Type);
+                    }
+                }
+
                 state.Current.ReturnValue = classInfo.CreateObject();
             }
+            else
+            {
+                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(classInfo.Type);
+            }
         }
 
         private static void HandleEndObject(ref ReadStack state)
         {
-            // Only allow dictionaries to be processed here if this is the DataExtensionProperty.
-            Debug.Assert(!state.Current.IsProcessingDictionary() || state.Current.JsonClassInfo.DataExtensionProperty == state.Current.JsonPropertyInfo);
-
-            if (state.Current.JsonClassInfo.ClassType == ClassType.Value)
-            {
-                // We should be in a converter, thus we must have bad JSON.
-                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.JsonPropertyInfo.RuntimePropertyType);
-            }
+            Debug.Assert(
+                state.Current.IsProcessingObject(ClassType.Object) ||
```

Comment:
The new condition for ClassType.Object doesn't appear to be needed since the `!state.Current.IsProcessingDictionary()` would cover that.